machine-api: update Azure Ultra Disks enhancement (#1060)
openshift/enhancements: enhancements/machine-api/azure-ultra-disks.md

  // +optional
  DataDisks []DataDisk `json:"dataDisks,omitempty"`
}

// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
// A Data Disk is a managed disk that's attached to a virtual machine to store application data.
// It differs from an OS Disk as it doesn't come with a pre-installed OS, and it cannot contain the boot volume.
// It is registered as SCSI drive and labeled with the chosen `lun`. e.g. for `lun: 0` the raw disk device will be available at `/dev/disk/azure/scsi1/lun0`.
//
// As the Data Disk disk device is attached raw to the virtual machine, it will need to be partitioned, formatted with a filesystem and mounted, in order for it to be usable.
// This can be done by creating a custom userdata Secret with custom Ignition configuration to achieve the desired initialization.
// At this stage the previously defined `lun` is to be used as the "device" key for referencing the raw disk device to be initialized.
// Once the custom userdata Secret has been created, it can be referenced in the Machine's `.providerSpec.userDataSecret`.

> // For further guidance and examples, please refer to the official OpenShift docs.

Reviewer: just curious if there is a link we can give here?

Author: There is a link but it's not specific for Data Disks or Azure, rather for AWS: https://docs.openshift.com/container-platform/4.10/post_installation_configuration/node-tasks.html#machine-node-custom-partition_post-install-node-tasks.
My intention is to have a chat with @jeana-redhat on how we can best proceed on the docs side and what to link from where :) If we decide on creating a specific doc section we could link it here and in the api godoc.

Reviewer 2: Cool - guess we will see if it is different enough to need its own heading or is AWS and Azure can use the same process with different details

Author: Sounds like a plan :)

Reviewer: +1, sounds great

### Implementation Details/Notes/Constraints

#### Initializing Data Disks 
While one of the goals for this enhancement is to provide automation for creating and attaching Azure Ultra Disks as Data Disks to Machines, there is one caveat.
Once an Azure instance is created and a Data Disk (in this case of type Ultra Disk) is attached to said instance, the Disk will be seen as a _raw_ block device by the OS.

This means that in order for the Data Disk to be usable by workloads it will require some form of initialization, which, given the high performing nature of Ultra Disks (and more generally of Data Disks) is best suited to be left to the cluster administrator.
This way they will be able to decide how to best slice the disk in partitions, while also choosing how to format them and autonomously deciding where to have them mounted.

> For this reason, guidance will be provided on how a cluster administrator can specify flexible configuration to achieve so with the help of a supporting example.

Reviewer: this makes it sound like the guidance will be provided somewhere else, but it looks like the following instructions are the guidance. if this is the guidance we are delivering, i would change this sentence to indicate that. for example:

```suggestion
For this reason, the following example is provided as guidance on how a cluster administrator can specify flexible configurations for their Data Disks:
```

Author: Makes sense, thanks. Included.
